pkg/block/blocktest/adapter.go

```diff
@@ -0,0 +1,311 @@
+				t.Errorf("Remove() tree diff = %s", diff)
+			}
+		})
+	}
+}
+
+func dumpPathTree(t testing.TB, ctx context.Context, adapter block.Adapter, qk block.QK) []string {
+	t.Helper()
+	tree := make([]string, 0)
+
+	uri, err := url.Parse(qk.Format())
+	require.NoError(t, err)
+
+	w, err := adapter.GetWalker(uri)
+	require.NoError(t, err)
+
+	walker := store.NewWrapper(w, uri)
+	require.NoError(t, err)
+
+	err = walker.Walk(ctx, block.WalkOptions{}, func(e block.ObjectStoreEntry) error {
+		_, p, _ := strings.Cut(e.Address, uri.String())
+		tree = append(tree, p)
+		return nil
+	})
+	if err != nil {
+		t.Fatalf("walking on '%s': %s", uri.String(), err)
+	}
+	sort.Strings(tree)
+	return tree
+}
+
+func testAdapterMultipartUpload(t *testing.T, adapter block.Adapter, storageNamespace string) {
+	ctx := context.Background()
+
+	cases := []struct {
+		name     string
+		path     string
+		partData []string
+	}{
+		{"simple", "abc", []string{"one ", "two ", "three"}},
+		{"nested", "foo/bar", []string{"one ", "two ", "three"}},
+	}
+
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			obj := block.ObjectPointer{
+				StorageNamespace: storageNamespace,
+				Identifier:       c.path,
+				IdentifierType:   block.IdentifierTypeRelative,
+			}
+			resp, err := adapter.CreateMultiPartUpload(ctx, obj, nil, block.CreateMultiPartUploadOpts{})
+			require.NoError(t, err)
+			parts := make([]block.MultipartPart, len(c.partData))
+			for partNumber, content := range c.partData {
+				partResp, err := adapter.UploadPart(ctx, obj, 0, strings.NewReader(content), resp.UploadID, partNumber)
+				require.NoError(t, err)
+				parts[partNumber].PartNumber = partNumber + 1
+				parts[partNumber].ETag = partResp.ETag
+			}
+			_, err = adapter.CompleteMultiPartUpload(ctx, obj, resp.UploadID, &block.MultipartUploadCompletion{
+				Part: parts,
+			})
+			require.NoError(t, err)
+
+			reader, err := adapter.Get(ctx, obj, 0)
+			require.NoError(t, err)
+
+			got, err := io.ReadAll(reader)
+			require.NoError(t, err)
+
+			expected := strings.Join(c.partData, "")
+			require.Equal(t, expected, string(got))
+		})
+	}
+}
+
+func testAdapterExists(t *testing.T, adapter block.Adapter, storageNamespace string) {
+	ctx := context.Background()
+	contents := "exists"
+	err := adapter.Put(ctx, block.ObjectPointer{
+		StorageNamespace: storageNamespace,
+		Identifier:       contents,
+		IdentifierType:   block.IdentifierTypeRelative,
+	}, 0, strings.NewReader(contents), block.PutOpts{})
+	require.NoError(t, err)
+
+	err = adapter.Put(ctx, block.ObjectPointer{
+		StorageNamespace: storageNamespace,
+		Identifier:       "nested/and/" + contents,
+		IdentifierType:   block.IdentifierTypeRelative,
+	}, 0, strings.NewReader(contents), block.PutOpts{})
+	require.NoError(t, err)
+
+	cases := []struct {
+		name   string
+		path   string
+		exists bool
+	}{
+		{"exists", "exists", true},
+		{"nested_exists", "nested/and/exists", true},
```

Comment:
```suggestion
		{"nested_exists", "nested/and/exists", true},
		{"nested_exists", "nested/and/../and/exists", falsee},
```
Supporting POSIX filesystems is _really_ nasty.